pkg/lib/cockpit-upload-helper.ts

```diff
@@ -0,0 +1,142 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2024 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import cockpit from 'cockpit';
+
+function debug(...args: unknown[]) {
+    if (window.debugging == 'all' || window.debugging?.includes('upload'))
+        console.debug('upload', ...args);
+}
+
+class UploadError extends Error {
+    name = 'UploadError';
+}
+
+/* XXX: these values are ~a decade old and were more or less copied from ssh
+ * (where they were last changed in 2007).  We might test alternatives...
+ */
+const BLOCK_SIZE = 16 << 10; // 16kiB
+const FLOW_WINDOW = 2 << 20; // 2MiB
+
+class Waiter {
+    unblock = () => { };
+    block(): Promise<void> {
+        return new Promise(resolve => { this.unblock = resolve });
+    }
+}
+
+export async function upload(
+    destination: string,
+    stream: ReadableStream,
+    progress?: (bytes_sent: number) => void,
+    signal?: AbortSignal,
+    options?: cockpit.JsonObject
+) {
+    let close_message = null as (cockpit.JsonObject | null);
+    let outstanding = 0; // for flow control
+    let delivered = 0; // for progress reporting
+
+    // This variable is the most important thing in this function.  The main
+    // upload loop will do work for as long as it can, and then it .block()s on
+    // the waiter until something changes (ack, close, abort, etc).  All of
+    // those things call .unblock() to resume the loop.
+    const event_waiter = new Waiter();
+
+    if (signal) {
+        signal.throwIfAborted(); // early exit
+        signal.addEventListener('abort', () => { event_waiter.unblock() });
```

Comment:
Totally not important, just for learning JS: You can drop the arrow and the block wrapper, this should work quite fine with directly invoking the method: `('abort', event_waiter.unblock)`.